src/main/java/net/neoforged/neoforge/internal/CommonModLoader.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.internal;
+
+import java.util.concurrent.Executor;
+import net.neoforged.api.distmarker.Dist;
+import net.neoforged.fml.ModList;
+import net.neoforged.fml.ModLoader;
+import net.neoforged.fml.ModWorkManager;
+import net.neoforged.fml.config.ConfigTracker;
+import net.neoforged.fml.config.ModConfig;
+import net.neoforged.fml.event.lifecycle.FMLClientSetupEvent;
+import net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent;
+import net.neoforged.fml.event.lifecycle.FMLDedicatedServerSetupEvent;
+import net.neoforged.fml.event.lifecycle.FMLLoadCompleteEvent;
+import net.neoforged.fml.event.lifecycle.InterModEnqueueEvent;
+import net.neoforged.fml.event.lifecycle.InterModProcessEvent;
+import net.neoforged.fml.loading.FMLEnvironment;
+import net.neoforged.fml.loading.FMLPaths;
+import net.neoforged.neoforge.network.registration.NetworkRegistry;
+import net.neoforged.neoforge.registries.GameData;
+import net.neoforged.neoforge.registries.RegistryManager;
+
+/**
+ * Internal class for handling the steps of mod loading that are common for client, data and server runs.
+ */
+public abstract class CommonModLoader {
+    private static boolean registriesLoaded = false;
+
+    public static boolean areRegistriesLoaded() {
+        return registriesLoaded;
+    }
+
+    protected static void begin(Runnable periodicTask) {
```

Comment:
I added a vague comment - the current separation is only an artifact of how we do part of the mod loading inside a reload listener (which introduces its own set of problems btw...). I don't really want to encode it because we have discussed a potential rework of the client mod init timings.